File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/application/usecases/TopUpConfirming.java
package it.giannibombelli.wsc2026.giftcard.application.usecases;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.application.UseCase;
import it.giannibombelli.wsc2026.giftcard.application.commands.ConfirmTopUp;
import it.giannibombelli.wsc2026.giftcard.domain.events.TopUpConfirmed;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import it.giannibombelli.wsc2026.giftcard.domain.ports.GiftCardRepository;

import static java.util.Objects.requireNonNull;

public class TopUpConfirming implements UseCase<ConfirmTopUp, TopUpConfirmed> {
    private final GiftCardRepository giftCardRepository;

    public TopUpConfirming(GiftCardRepository giftCardRepository) {
        this.giftCardRepository = requireNonNull(giftCardRepository);
    }

    @Override
    public TopUpConfirmed invoke(ConfirmTopUp cmd) {
        Require.requireArgument(cmd, "command");
        GiftCardId giftCardId = cmd.aggregateId();

        GiftCard giftCard = giftCardRepository.findById(giftCardId)
            .orElseThrow(() -> new IllegalStateException("GiftCard not found for top-up confirmation: " + giftCardId));

        TopUpConfirmed confirmed = giftCard.confirmTopUp(cmd.amount());

        giftCardRepository.save(giftCard);
        return confirmed;
    }
}

File: typescript/src/giftcard/application/usecases/topUpConfirming.ts
// Caso d'uso per la conferma della ricarica di una gift card.

import type { UseCase } from '@/common/application/usecase.js';
import type { TopUpConfirmed } from '../../domain/events/topUpConfirmed.js';
import { GiftCardId } from '../../domain/giftcard/giftCardId.js';
import type { GiftCardRepository } from '../../domain/ports/giftCardRepository.js';
import { ConfirmTopUp } from '../commands/confirmTopUp.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class TopUpConfirming implements UseCase<ConfirmTopUp, TopUpConfirmed> {
  private readonly _giftCardRepository: GiftCardRepository;

  constructor(giftCardRepository: GiftCardRepository) {
    requireDependency(giftCardRepository, "giftCardRepository");
    this._giftCardRepository = giftCardRepository;
  }

  invoke(command: ConfirmTopUp): TopUpConfirmed {
    requireArgument(command, 'command');

    const giftCardId: GiftCardId = command.aggregateId;
    const giftCard = this._giftCardRepository.findById(giftCardId);
    if (giftCard === null) {
      throw new Error(`GiftCard not found for top-up confirmation: ${giftCardId.value}`);
    }

    const confirmed = giftCard.confirmTopUp(command.amount);

    this._giftCardRepository.save(giftCard);
    return confirmed;
  }
}
